kratos/utilities/geometry_utilities/nearest_point_utilities.cpp

```diff
@@ -0,0 +1,208 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+//  Main authors:    Pooyan Dadvand
+//                   Vicente Mataix Ferrandiz
+//
+
+// System includes
+
+// External includes
+
+// Project includes
+#include "utilities/math_utils.h"
+#include "utilities/geometry_utilities/nearest_point_utilities.h"
+
+namespace Kratos
+{
+
+Point NearestPointUtilities::LineNearestPoint(
+    const array_1d<double, 3>& rPoint,
+    const array_1d<double, 3>& rLinePointA,
+    const array_1d<double, 3>& rLinePointB
+    )
+{
+    // Project point globally into the line
+    const array_1d<double, 3> ab = rLinePointB - rLinePointA;
+    const double factor = (inner_prod(rLinePointB, rPoint) - inner_prod(rLinePointA, rPoint) - inner_prod(rLinePointB, rLinePointA) + inner_prod(rLinePointA, rLinePointA)) / inner_prod(ab, ab);
+    Point result(rLinePointA + factor * ab);
+
+    // Compute square_length of the line
+    double lx = rLinePointA[0] - rLinePointB[0];
+    double ly = rLinePointA[1] - rLinePointB[1];
+    double lz = rLinePointA[2] - rLinePointB[2];
+    double square_length = lx * lx + ly * ly + lz * lz;
+
+    // Compute square_length of the projection (A)
+    lx = result[0] - rLinePointA[0];
+    ly = result[1] - rLinePointA[1];
+    lz = result[2] - rLinePointA[2];
+    double square_length_1 = lx * lx + ly * ly + lz * lz;
+
+    // Compute square_length of the projection (B)
+    lx = result[0] - rLinePointB[0];
+    ly = result[1] - rLinePointB[1];
+    lz = result[2] - rLinePointB[2];
+    double square_length_2 = lx * lx + ly * ly + lz * lz;
+
+    // Project point locally into the line
+    array_1d<double,3> projected_local;
+    const double tolerance = 1e-14; // Tolerance
```

Comment:
I would take the tolerance as an optional argument.

Reply:
Okay